### PR TITLE
Block mover button: do not show focus styles on pointer interactions

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Tweak mover button styles in both the block toolbar and the list view, so that the button's focus styles are not shown as the results of a user click.

## Why?

More polished button styles when interacting with a pointer.

## How?

By using the [`:focus-visible` CSS pseudoselector](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible), which lets the browser determine via heuristics if the focus should be made evident on the element — in practice, it allows for focus styles to be shown only on keyboard interaction (and not as a result of pointer clicks).

 - moved the block mover button styles to `packages/block-editor/src/components/block-mover/style.scss`
- I still had to leave those same styles in place in `packages/components/src/toolbar/style.scss` because they are applied to the rest of toolbar button (using `:focus` instead of `:focus-visible`)
- I had to add an override in the block mover button styles to remove the `:focus` styles applied by the block toolbar
- I had to leave some minor overrides for the size/position of the `::before` pseudo element in the list view

_Note: the correct way of making this change would be by changing styles directly in the `Button` component. This PR instead resorts to applying style overrides directly in the block editor — this is a sub-optimal approach, which introduces technical debt and fragmentation, making it actually harder to work on the `Button` component in the future. This approach is therefore not reccommended._

## Testing Instructions

Block mover buttons in the toolbar:
- on desktop viewport sizes:
  1. Select a block
  2. In the toolbar, click on the up/down block mover buttons
  3. Notice how the focus ring doesn't show
  4. Navigate the toolbar with the keyboard (tab & and arrow keys)
  5. Notice how the mover buttons show a focus rings when focus via keyboard interaction
- Resize the viewport to mobile size, the toolbar should move to a fixed positoin on top of the editor
  1. Repeat steps 1 to 5 above

Block mover buttons in the tree list:
- enable block mover buttons (I did it by [setting this `hasRenderedMovers` variable to `true`](https://github.com/WordPress/gutenberg/blob/820d8cfb1366354867cd212a51c9388a971888ad/packages/block-editor/src/components/list-view/block.js#L137))
- Repeat steps 2 to 5 above

### Before (`trunk`): focus ring appears on both pointer and keyboard interactions

Block toolbar:

https://user-images.githubusercontent.com/1083581/196880269-345440ab-16d0-46c4-bd3a-a4edafd5bbd0.mp4

List view:

https://user-images.githubusercontent.com/1083581/196880283-7ce52515-de7c-42b6-b8db-1ccc59c9b369.mp4

### After (this PR): focus ring only shows on keyboard interactions

Block toolbar:

https://user-images.githubusercontent.com/1083581/196880289-7c2d888e-5fcd-4e89-ab10-dd7d4f5a5063.mp4

List view:

https://user-images.githubusercontent.com/1083581/196880300-b7bceb93-cc2b-44c4-b113-904dcfb726b8.mp4

### CODE REVIEW

**Comment by talldan** [comment]
> Tweak mover button styles

This PR seems to apply to all toolbar button styles, which does make things more consistent. Is that on purpose?

**Comment by youknowriad** [comment]
What's the guideline for the new focus-visible pseudo selectors? Should we use them consistently for all our focus styles?

Also, it's actually already supported in all the browsers we do support, so technically we could avoid the "supports" fallback or you're just leaning on the safe side here.

**Comment by ciampo** [comment]
> > Tweak mover button styles
> 
> This PR seems to apply to all toolbar button styles, which does make things more consistent.

Great catch, I had actually meant to update the title / description but clearly forgot.

> Is that on purpose?

It is for now, but it definitely needs approval from folks like @mtias / @jasmussen — it's hard to draw the line of where the new focus styles should apply: just the block movers? All buttons in the toolbar? What about the dropdown menus originating from the toolbar? And what about the list view, and the remaining buttons in the list view?

The more we spread out these styles, the more consistent the UI looks — but that means introducing even more style overrides and more technical debt.

**Comment by ciampo** [comment]
> What's the guideline for the new focus-visible pseudo selectors? Should we use them consistently for all our focus styles?

I'd love to, but that would mean to refactor the `Button` component. Problem is, there is a high amount of style overrides for the `Button` component — including focus styles — throughout the repository (for example, the buttons in the toolbar completely disable `Button`'s styles and implement new focus styles via the `::before` pseudo-element). Which would make this refactor a pretty large task, and a complex one to get right without introducing regressions.

I think that the spirit of Matias' request was to just apply this change to the block movers.

> Also, it's actually already supported in all the browsers we do support, so technically we could avoid the "supports" fallback or you're just leaning on the safe side here.

From what I know, se support the latest 2 version of evergreen browsers — support for `:focus-visible` was introduced in Safari 15.4, which means we'd need fallback styles to support Safari 15-15.3.

**Comment by ciampo** [comment]
> I think that the spirit of Matias' request was to just apply this change to the block movers.

With this in mind, I've [pushed a change](https://github.com/WordPress/gutenberg/pull/45126/commits/41f76b223751549cf4abe2860570c9e68b0454df) that apply the new `:focus-visible` styles only to block mover buttons. The rest of the buttons in the toolbar keep using the same focus styles as on `trunk` (ie. they show a ring when clicking)

**Comment by youknowriad** [comment]
@ciampo I think the last version of Safari is Safari 16 but it probably depends on the OS.

**Comment by ciampo** [comment]
> @ciampo I think the last version of Safari is Safari 16 but it probably depends on the OS.

I think I know where the misunderstanding lies — I was interpreting it the word "version" as "major version", and therefore I thought that we had to support Safari 15 and 16 😅 

But looking at browserlist's website, the current "last 2 versions" for Safari are `16.0` and `15.6`

Screenshot 2022-10-20 at 11 45 45

and even checking for "> 1%" usage doesn't show any Safari version before `15.4`

Screenshot 2022-10-20 at 11 46 04


Which means that I can safely remove the fallback. Sorry for the confusion!

File: packages/components/src/toolbar/style.scss
Comment (mirka) [inline packages/components/src/toolbar/style.scss]:
```suggestion
		// Focus style for toolbar buttons, except block movers.
```

File: packages/components/src/toolbar/style.scss
Comment (mirka) [inline packages/components/src/toolbar/style.scss]: Do you prefer that these hacks live here instead of `packages/block-editor/src/components/block-mover/style.scss`?

I think I prefer moving it to `block-mover` unless it's likely that we're defaulting to focus-visible for the rest of the toolbar buttons soon. `@wordpress/components` shouldn't contain downstream overrides, even if that means dirtier code downstream (e.g. additional specificity hacks).

Comment (ciampo) [reply packages/components/src/toolbar/style.scss]: I attempted to move styles to `packages/block-editor/src/components/block-mover/style.scss` as you suggested:

- I still had to leave those same styles in place in `packages/components/src/toolbar/style.scss` because they are applied to the rest of toolbar button (using `:focus`)
- I had to add an override in the block mover button styles to remove the `:focus` styles applied by the block toolbar
- I had to leave some minor overrides for the size/position of the `::before` pseudo element in the list view


https://github.com/WordPress/gutenberg/pull/45126/commits/6c17fae8e9a2a169f106c0ab0a1d60b6ba531712 — WDYT ?

File: packages/block-editor/src/components/list-view/block.js
Comment (ciampo) [inline packages/block-editor/src/components/list-view/block.js]: I didn't intend to commit this line, but we can keep it for now to make it easier to review the block mover button in the list view. I will remove this line before merging

File: packages/components/src/toolbar/style.scss
Comment (mirka) [reply packages/components/src/toolbar/style.scss]: Whoa, talk about leaving things cleaner than you found them 👏 Great refactor!

**Review by mirka — APPROVED** [review]
I noticed this issue where the focus ring colors aren't properly inverted when the background is blue:

Focus style when background is white
Incorrect, non-inverted focus style when background is blue

But this is a completely separate and pre-existing issue. Let's merge this one 🚢

**Pull request merged**
